Xamarin.Forms.Xaml.UnitTests/DataTemplateExtension.xaml.cs

```diff
@@ -0,0 +1,33 @@
+﻿using NUnit.Framework;
+
+using Xamarin.Forms.Core.UnitTests;
+
+namespace Xamarin.Forms.Xaml.UnitTests
+{
+	public partial class DataTemplateExtension : ContentPage
+	{
+		public DataTemplateExtension() => InitializeComponent();
+		public DataTemplateExtension(bool useCompiledXaml)
+		{
+			//this stub will be replaced at compile time
+		}
+
+		[TestFixture]
+		class Tests
+		{
+			[SetUp]
+			public void Setup() => Device.PlatformServices = new MockPlatformServices();
+			[TearDown] public void TearDown() => Device.PlatformServices = null;
+
+			[TestCase(true), TestCase(false)]
+			public void DataTemplateExtension(bool useCompiledXaml)
```

Comment:
How does this test whether the markup extension was the compile one or the dynamic one?

Reply:
this assumes that, in the case when `useCompiledXaml` is `true`, the `ctor` in
```csharp
var layout = new DataTemplateExtension(useCompiledXaml);
```
will return the compiled version of the Xaml.

And this test that the compiled version behave as it should.

There's no easy way to test that the markup as actually been compiled, as it has no side effects, and leave lo artifacts. But let me show you that it actually works:

```il
	IL_005a:  nop 
	IL_005b:  ldtoken Xamarin.Forms.Xaml.UnitTests.DataTemplateExtension
	IL_0060:  call class [mscorlib]System.Type class [mscorlib]System.Type::GetTypeFromHandle(valuetype [mscorlib]System.RuntimeTypeHandle)
	IL_0065:  newobj instance void class [Xamarin.Forms.Core]Xamarin.Forms.DataTemplate::'.ctor'(class [mscorlib]System.Type)
	IL_006a:  stloc.0 
```

and then, further down the `il`

```il
	IL_0080:  ldloc.1 
	IL_0081:  ldsfld class [Xamarin.Forms.Core]Xamarin.Forms.BindableProperty [Xamarin.Forms.Core]Xamarin.Forms.ShellContent::ContentTemplateProperty
	IL_0086:  ldloc.0 
	IL_0087:  callvirt instance void class [Xamarin.Forms.Core]Xamarin.Forms.BindableObject::SetValue(class [Xamarin.Forms.Core]Xamarin.Forms.BindableProperty, object)
```

(local `V_1` is the `ShellContent` instance)

so this does shows that the markup extension is not evaluated at runtime, but actually compiled to the equiv code of
```csharp
new DataTemplate(typeof(Foo));
```